File: android/src/main/kotlin/net/niuxiaoer/flutter_gromore/event/AdEventHandler.kt
package net.niuxiaoer.flutter_gromore.event

import android.util.Log
import io.flutter.plugin.common.EventChannel

class AdEventHandler: EventChannel.StreamHandler {
    private val TAG: String = this::class.java.simpleName
    private var eventSink: EventChannel.EventSink? = null

    companion object {
        // 单例
        fun getInstance() = InstanceHelper.instance
    }

    object InstanceHelper {
        val instance = AdEventHandler()
    }

    fun sendEvent(adEvent: AdEvent) {
        eventSink?.success(adEvent.toMap())
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        Log.d(TAG, "onListen")
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        Log.d(TAG, "onCancel")
        eventSink = null
    }
}

File: ios/Classes/event/AdEventHandler.swift
//
//  AdEventHandler.swift
//  flutter_gromore
//
//  Created by Anand on 2022/6/2.
//

class AdEventHandler: NSObject, FlutterStreamHandler {
    static let instance: AdEventHandler = AdEventHandler()
    private var eventSink: FlutterEventSink?
    
    func sendEvent(_ event: AdEvent) {
        if let sink = eventSink {
            sink(event.toMap())
        }
    }
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}
